File: qt/network/gen_qhttpmultipart.cpp
// +build ignore

#include <QByteArray>
#include <QHttpMultiPart>
#include <QHttpPart>
#include <QIODevice>
#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVariant>
#include <qhttpmultipart.h>
#include "gen_qhttpmultipart.h"

QHttpPart* QHttpPart_new() {
	return new QHttpPart();
}

QHttpPart* QHttpPart_new2(QHttpPart* other) {
	return new QHttpPart(*other);
}

void QHttpPart_OperatorAssign(QHttpPart* self, QHttpPart* other) {
	self->operator=(*other);
}

void QHttpPart_Swap(QHttpPart* self, QHttpPart* other) {
	self->swap(*other);
}

bool QHttpPart_OperatorEqual(const QHttpPart* self, QHttpPart* other) {
	return (*self == *other);
}

bool QHttpPart_OperatorNotEqual(const QHttpPart* self, QHttpPart* other) {
	return (*self != *other);
}

void QHttpPart_SetHeader(QHttpPart* self, int header, QVariant* value) {
	self->setHeader(static_cast<QNetworkRequest::KnownHeaders>(header), *value);
}

void QHttpPart_SetRawHeader(QHttpPart* self, struct miqt_string headerName, struct miqt_string headerValue) {
	QByteArray headerName_QByteArray(headerName.data, headerName.len);
	QByteArray headerValue_QByteArray(headerValue.data, headerValue.len);
	self->setRawHeader(headerName_QByteArray, headerValue_QByteArray);
}

void QHttpPart_SetBody(QHttpPart* self, struct miqt_string body) {
	QByteArray body_QByteArray(body.data, body.len);
	self->setBody(body_QByteArray);
}

void QHttpPart_SetBodyDevice(QHttpPart* self, QIODevice* device) {
	self->setBodyDevice(device);
}

void QHttpPart_Delete(QHttpPart* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QHttpPart*>( self );
	} else {
		delete self;
	}
}

class MiqtVirtualQHttpMultiPart : public virtual QHttpMultiPart {
public:

	MiqtVirtualQHttpMultiPart(): QHttpMultiPart() {};
	MiqtVirtualQHttpMultiPart(ContentType contentType): QHttpMultiPart(contentType) {};
	MiqtVirtualQHttpMultiPart(QObject* parent): QHttpMultiPart(parent) {};
	MiqtVirtualQHttpMultiPart(ContentType contentType, QObject* parent): QHttpMultiPart(contentType, parent) {};

	virtual ~MiqtVirtualQHttpMultiPart() = default;

	// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;

	// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
		if (handle__MetaObject == 0) {
			return QHttpMultiPart::metaObject();
		}
		

		QMetaObject* callback_return_value = miqt_exec_callback_QHttpMultiPart_MetaObject(const_cast<MiqtVirtualQHttpMultiPart*>(this), handle__MetaObject);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QHttpMultiPart::metaObject();

	}

	// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;

	// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
		if (handle__Metacast == 0) {
			return QHttpMultiPart::qt_metacast(param1);
		}
		
		const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QHttpMultiPart_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QHttpMultiPart::qt_metacast(param1);

	}

};

QHttpMultiPart* QHttpMultiPart_new() {
	return new MiqtVirtualQHttpMultiPart();
}

QHttpMultiPart* QHttpMultiPart_new2(ContentType contentType) {
	return new MiqtVirtualQHttpMultiPart(contentType);
}

QHttpMultiPart* QHttpMultiPart_new3(QObject* parent) {
	return new MiqtVirtualQHttpMultiPart(parent);
}

QHttpMultiPart* QHttpMultiPart_new4(ContentType contentType, QObject* parent) {
	return new MiqtVirtualQHttpMultiPart(contentType, parent);
}

void QHttpMultiPart_virtbase(QHttpMultiPart* src, QObject** outptr_QObject) {
	*outptr_QObject = static_cast<QObject*>(src);
}

QMetaObject* QHttpMultiPart_MetaObject(const QHttpMultiPart* self) {
	return (QMetaObject*) self->metaObject();
}

void* QHttpMultiPart_Metacast(QHttpMultiPart* self, const char* param1) {
	return self->qt_metacast(param1);
}

struct miqt_string QHttpMultiPart_Tr(const char* s) {
	QString _ret = QHttpMultiPart::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

void QHttpMultiPart_Append(QHttpMultiPart* self, QHttpPart* httpPart) {
	self->append(*httpPart);
}

void QHttpMultiPart_SetContentType(QHttpMultiPart* self, ContentType contentType) {
	self->setContentType(contentType);
}

struct miqt_string QHttpMultiPart_Boundary(const QHttpMultiPart* self) {
	QByteArray _qb = self->boundary();
	struct miqt_string _ms;
	_ms.len = _qb.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _qb.data(), _ms.len);
	return _ms;
}

void QHttpMultiPart_SetBoundary(QHttpMultiPart* self, struct miqt_string boundary) {
	QByteArray boundary_QByteArray(boundary.data, boundary.len);
	self->setBoundary(boundary_QByteArray);
}

struct miqt_string QHttpMultiPart_Tr2(const char* s, const char* c) {
	QString _ret = QHttpMultiPart::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QHttpMultiPart_Tr3(const char* s, const char* c, int n) {
	QString _ret = QHttpMultiPart::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

void QHttpMultiPart_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQHttpMultiPart*>( (QHttpMultiPart*)(self) )->handle__MetaObject = slot;
}

QMetaObject* QHttpMultiPart_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQHttpMultiPart*)(self) )->virtualbase_MetaObject();
}

void QHttpMultiPart_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQHttpMultiPart*>( (QHttpMultiPart*)(self) )->handle__Metacast = slot;
}

void* QHttpMultiPart_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQHttpMultiPart*)(self) )->virtualbase_Metacast(param1);
}

void QHttpMultiPart_Delete(QHttpMultiPart* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQHttpMultiPart*>( self );
	} else {
		delete self;
	}
}

File: qt/network/gen_qhttpmultipart.go
package network

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QHttpMultiPart__ContentType int

const (
	QHttpMultiPart__MixedType       QHttpMultiPart__ContentType = 0
	QHttpMultiPart__RelatedType     QHttpMultiPart__ContentType = 1
	QHttpMultiPart__FormDataType    QHttpMultiPart__ContentType = 2
	QHttpMultiPart__AlternativeType QHttpMultiPart__ContentType = 3
)

type QHttpPart struct {
	h          uintptr
	isSubclass bool
}

// NewQHttpPart constructs a new QHttpPart object.
func NewQHttpPart() *QHttpPart {
	g := newQHttpPart(QHttpPart_new())
	g.isSubclass = true
	return g
}

// NewQHttpPart2 constructs a new QHttpPart object.
func NewQHttpPart2(other *QHttpPart) *QHttpPart {
	g := newQHttpPart(QHttpPart_new2(other.cPointer()))
	g.isSubclass = true
	return g
}

func (this *QHttpPart) OperatorAssign(other *QHttpPart) {
	QHttpPart_OperatorAssign(this.h, other.cPointer())
}

func (this *QHttpPart) Swap(other *QHttpPart) {
	QHttpPart_Swap(this.h, other.cPointer())
}

func (this *QHttpPart) OperatorEqual(other *QHttpPart) bool {
	return (bool)(QHttpPart_OperatorEqual(this.h, other.cPointer()))
}

func (this *QHttpPart) OperatorNotEqual(other *QHttpPart) bool {
	return (bool)(QHttpPart_OperatorNotEqual(this.h, other.cPointer()))
}

func (this *QHttpPart) SetHeader(header QNetworkRequest__KnownHeaders, value *qt.QVariant) {
	QHttpPart_SetHeader(this.h, (int)(header), (*QVariant)(value.UnsafePointer()))
}

func (this *QHttpPart) SetRawHeader(headerName []byte, headerValue []byte) {
	headerName_alias := struct_miqt_string{}
	headerName_alias.data = (char)(unsafe.Pointer(&headerName[0]))
	headerName_alias.len = size_t(len(headerName))
	headerValue_alias := struct_miqt_string{}
	headerValue_alias.data = (char)(unsafe.Pointer(&headerValue[0]))
	headerValue_alias.len = size_t(len(headerValue))
	QHttpPart_SetRawHeader(this.h, headerName_alias, headerValue_alias)
}

func (this *QHttpPart) SetBody(body []byte) {
	body_alias := struct_miqt_string{}
	body_alias.data = (char)(unsafe.Pointer(&body[0]))
	body_alias.len = size_t(len(body))
	QHttpPart_SetBody(this.h, body_alias)
}

func (this *QHttpPart) SetBodyDevice(device *qt.QIODevice) {
	QHttpPart_SetBodyDevice(this.h, (*QIODevice)(device.UnsafePointer()))
}

type QHttpMultiPart struct {
	h          uintptr
	isSubclass bool
}

// NewQHttpMultiPart constructs a new QHttpMultiPart object.
func NewQHttpMultiPart() *QHttpMultiPart {
	g := newQHttpMultiPart(QHttpMultiPart_new())
	g.isSubclass = true
	return g
}

// NewQHttpMultiPart2 constructs a new QHttpMultiPart object.
func NewQHttpMultiPart2(contentType ContentType) *QHttpMultiPart {
	g := newQHttpMultiPart(QHttpMultiPart_new2(contentType))
	g.isSubclass = true
	return g
}

// NewQHttpMultiPart3 constructs a new QHttpMultiPart object.
func NewQHttpMultiPart3(parent *qt.QObject) *QHttpMultiPart {
	g := newQHttpMultiPart(QHttpMultiPart_new3((*QObject)(parent.UnsafePointer())))
	g.isSubclass = true
	return g
}

// NewQHttpMultiPart4 constructs a new QHttpMultiPart object.
func NewQHttpMultiPart4(contentType ContentType, parent *qt.QObject) *QHttpMultiPart {
	g := newQHttpMultiPart(QHttpMultiPart_new4(contentType, (*QObject)(parent.UnsafePointer())))
	g.isSubclass = true
	return g
}

func (this *QHttpMultiPart) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QHttpMultiPart_MetaObject(this.h)))
}

func (this *QHttpMultiPart) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QHttpMultiPart_Metacast(this.h, param1_Cstring))
}

func QHttpMultiPart_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QHttpMultiPart_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QHttpMultiPart) Append(httpPart *QHttpPart) {
	QHttpMultiPart_Append(this.h, httpPart.cPointer())
}

func (this *QHttpMultiPart) SetContentType(contentType ContentType) {
	QHttpMultiPart_SetContentType(this.h, contentType)
}

func (this *QHttpMultiPart) Boundary() []byte {
	var _bytearray struct_miqt_string = QHttpMultiPart_Boundary(this.h)
	_ret := GoBytes(unsafe.Pointer(_bytearray.data), int(int64(_bytearray.len)))
	free(unsafe.Pointer(_bytearray.data))
	return _ret
}

func (this *QHttpMultiPart) SetBoundary(boundary []byte) {
	boundary_alias := struct_miqt_string{}
	boundary_alias.data = (char)(unsafe.Pointer(&boundary[0]))
	boundary_alias.len = size_t(len(boundary))
	QHttpMultiPart_SetBoundary(this.h, boundary_alias)
}

func QHttpMultiPart_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QHttpMultiPart_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QHttpMultiPart_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QHttpMultiPart_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QHttpMultiPart) callVirtualBase_MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QHttpMultiPart_virtualbase_MetaObject(unsafe.Pointer(this.h))))
}

func (this *QHttpMultiPart) OnMetaObject(slot func(super func() *qt.QMetaObject) *qt.QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QHttpMultiPart_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QHttpMultiPart_MetaObject
func miqt_exec_callback_QHttpMultiPart_MetaObject(self QHttpMultiPart, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *qt.QMetaObject) *qt.QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QHttpMultiPart{h: self}).callVirtualBase_MetaObject)

	return (*QMetaObject)(virtualReturn.UnsafePointer())
}

func (this *QHttpMultiPart) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QHttpMultiPart_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QHttpMultiPart) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QHttpMultiPart_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QHttpMultiPart_Metacast
func miqt_exec_callback_QHttpMultiPart_Metacast(self QHttpMultiPart, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}
	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)
	virtualReturn := gofunc((&QHttpMultiPart{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
